src/mds/MDCache.cc

```diff
@@ -9811,7 +9811,25 @@ void MDCache::request_forward(const MDRequestRef& mdr, mds_rank_t who, int port)
 void MDCache::dispatch_request(const MDRequestRef& mdr)
 {
   if (mdr->dead) {
-    dout(20) << __func__ << ": dead " << *mdr << dendl;
+    dout(10) << __func__ << ": dead " << *mdr << dendl;
+    //if the mdr is a "batch_op" and it has followers, pick a follower as
+    //the new "head of the batch ops" and go on processing the new one.
+    if (mdr->client_request && mdr->is_batch_head()) {
```

Comment:
I understand why this works but it is weird not to select a new batch head in `::request_cleanup`. I would have preferred that but I think this PR needs to be merged urgently. Maybe in a follow-up cleanup?

Reply:
Does it really need ?

```
 2079 void Server::respond_to_request(const MDRequestRef& mdr, int r)
 2080 {
 2081   mdr->result = r;
 2082   if (mdr->client_request) {
 2083     if (mdr->is_batch_head()) {
 2084       dout(20) << __func__ << ": batch head " << *mdr << dendl;
 2085       mdr->release_batch_op()->respond(r);                                                                                                                                                                                                              
 2086     } else {
 2087      reply_client_request(mdr, make_message<MClientReply>(*mdr->client_request, r));
 2088     }
 2089   } else if (mdr->internal_op > -1) { 
 2090     dout(10) << __func__ << ": completing with result " << cpp_strerror(r) << " on internal " << *mdr << dendl;
 2091     auto c = mdr->internal_op_finish;   
 2092     if (!c)
 2093       ceph_abort_msg("trying to respond to internal op without finisher");
 2094     mdcache->request_finish(mdr);
 2095     c->complete(r);
 2096   }
 2097 }
```

We can see in Line#2085 it will cleanup all the batch requests when the batch head finishes.

Reply:
Let me have a check carefully and try to improve the code.

Reply:
> ```
> 
> ```
> 
> 
>     
>   
> 
> We can see in Line#2085 it will cleanup all the batch requests when the batch head finishes.

In this case, `Server::respond_to_request` is not called because the request was killed (that's why `dead == true`). This logic for batch_head handling in `::respond_to_request` and now ` MDCache::dispatch_request` should be uniformly handled in a single location: `MDCache::request_cleanup`.

Reply:
> @batrick please confirm that the `if (dead)` check only applies to internal requests, and can be moved after the client and peer conditional dispatches

It is also handled for client requests. For example:

- getattr waits on lock in `Locker::acquire_locks`
- client running getattr is evicted, all its requests are killed and cleaned up (`dead = true;`)
- Locker resumes the request when the lock is available, it trips on `if (dead)` in dispatch and exits

Reply:
The `if (dead)` check was added as part of our quiesce changes. Without that code, the request would get into

```
  if (mdr->client_request) {
    mds->server->dispatch_client_request(mdr);
```

and get processed, handling the batch head case. I don't see how `dead` would affect that flow, but apparently, it wasn't ever a problem until now, was it?

Reply:
The finding above is by @lxbsz

Reply:
Ah. I think it's even more important to unify the handling here because now we have three places that rectify the batch_head request dying/killed/finishing:

- `MDCache::dispatch_request`
- `Server::dispatch_client_request`
- `Server::respond_to_request`

This should all be moved to `MDCache::request_cleanup`.

Reply:
right, this is what @lxbsz had also figured out. So my question then is, shouldn't we just move that new `if(dead)` check into the else?

Reply:
No, I think we need to make a better effort unify handling of requests in the MDS. A dead request should not be dispatched no matter its type.

Reply:
> @batrick But in the case when the sessions are killed and then all the corresponding requests will be marked killed too. Is that fine if we just remove the queued requests from the finisher queue directly ? If not then we cannot avoid this IMO, because the killed requests maybe still in the finisher queue and not dispatched yet.

I understand. I am probably being confusing. It's okay that a dead request is re-dispatched through `MDCache::dispatch_request` **but** if it's marked `dead`, do nothing else and `return;`.

> But we can just choose a batch head directly when the a batch head request is being killed instead of doing this when the request is being dispatched.

Yes!

Reply:
Let me try to improve it. Thanks @batrick @leonid-s-usov